### DISCUSSION TITLE
Cannot make `recover_with(via_parser(…))` work

### DISCUSSION THREAD

**Discussion by soywod** [zesterer/chumsky#609] [Q&A] [answered]
Sorry if the question sounds stupid, but I cannot figure out how the `recover_with` and `via_parser` work (using `alpha.6`). I have the following:

```rust
let parser =                                                                          
    just::<_, _, extra::Err<Rich<char>>>("abc").recover_with(via_parser(just("abd")));
assert_eq!(parser.parse("abd").into_result().unwrap(), "abd");                        
```

which outputs:

```text
called `Result::unwrap()` on an `Err` value: [found 'd' at 2..3 expected ''c'']
```

I naively expect `abc` not being parsed to be recovered by `abd`, what am I missing?

---

To give more context, I try to parse either a parser 1 or a parser 2, and both contain similar keywords. Knowing that chumsky is a PEG, a `choice((…))` does not work as the beginning of the parser 1 can be partially parsed by the parser 2 and vice versa, which leads to parsing errors. Is error recovery the right approach for such issue?

**Comment by zesterer** [comment, marked as answer]
The result here is both expected and correct: parser recovery doesn't mean that the error in the first branch (`just("abc")`) gets swallowed, it just means that an error gets emitted and then parsing *continues*, using the recovery strategy to 'fill the whole', as it were.

If you check out the docs for [`ParseResult`](https://docs.rs/chumsky/1.0.0-alpha.6/chumsky/struct.ParseResult.html), you'll see that it's possible for a parser to both produce a valid output *and* emit errors at the same time. This is what is happening here. `recover_with(via_parser(...))` should only be used to recover from errors.

If both patterns are considered valid syntax, then `choice`/`or` are the thing to use (both do the same thing internally).
